src/olmo_core/data/source_mixture.py

```diff
@@ -0,0 +1,307 @@
+    """
+    A class to hold intermediate selection details for a mixture source.
+    """
+
+    config: SourceMixtureConfig
+    population: int
+    num_selected: int
+
+    def for_table(self, max_tokens: int) -> Dict:
+        return {
+            "source_name": self.config.source_name,
+            "source_population": f"{self.population:.2e}",
+            "num_sampled": f"{self.num_selected:.2e}",
+            "target_ratio": self.config.target_ratio,
+            "max_repetion_ratio": self.config.max_repetition_ratio,
+            "max_source_fraction": self.config.max_source_fraction,
+            "observed_source_ratio": f"{(self.num_selected / self.population):.4}",
+            "observed_global_ratio": f"{(self.num_selected / max_tokens):.4}",
+        }
+
+
+@dataclass
+class SourcePathTokens:
+    path: PathOrStr
+    tokens: int
+
+
+@dataclass
+class SourceMixtureOutcome:
+    name: str
+    path_tokens: List[SourcePathTokens]
+
+
+@dataclass
+class SourceMixtureDataset:
+    """
+    A dataset consisting of a fractionalized mixture of data sources.
+    """
+
+    sources: List[SourceMixtureOutcome]
+
+    def to_index(self) -> Dict[Tuple[str, int], int]:
+        """
+        Convert the dataset to an indexed array of dict((int, path), int).
+        """
+        return {
+            (str(outcome.path), idx): outcome.tokens
+            for idx, outcome in enumerate(
+                list(chain.from_iterable([outcome.path_tokens for outcome in self.sources]))
+            )
+        }
+
+    def to_paths(self) -> List[PathOrStr]:
+        """
+        Convert the dataset to a list of paths while maintaining stable ordering.
+        """
+        return [
+            item.path
+            for item in list(chain.from_iterable([outcome.path_tokens for outcome in self.sources]))
+        ]
+
+
+@dataclass
+class SourceMixtureDatasetConfig(Config):
+    """
+    A configuration class for building a dataset from a fractionalized mixture of sources.
+    """
+
+    max_tokens: int
+    source_configs: List[SourceMixtureConfig]
+    sequence_length: int
+    dtype: NumpyDatasetDType
+    processes: int = 1
+    seed: int = 42
+
+    def validate(self):
+        if self.max_tokens <= 0:
+            raise OLMoConfigurationError("max_tokens must be > 0")
+
+        if not self.source_configs:
+            raise OLMoConfigurationError("source_configs must not be empty")
+
+        if (total := sum([source.target_ratio for source in self.source_configs])) != 1.0:
+            raise OLMoConfigurationError(f"target_ratios must sum to 1, got {total}")
+
+    def build(self) -> SourceMixtureDataset:
+        self.validate()
+        random.seed(self.seed)
+        available_tokens_by_source: Dict[str, int] = {}
+
+        print("---------------------------------------------------------")
+        print("Generating a source mixture from configurations:")
+        for source_config in self.source_configs:
+            pprint(source_config)
```

Comment:
Our logging will actually pretty print config objects.
```suggestion
            log.info(source_config)
```